cmd/eval-dev-quality/cmd/evaluate.go

```diff
@@ -421,53 +409,67 @@ func (command *Evaluate) Initialize(args []string) (evaluationContext *evaluate.
 				}
 				serviceShutdown = append(serviceShutdown, shutdown)
 			}
+		}
 
-			// Check if a provider has the ability to pull models and do so if necessary.
-			if puller, ok := p.(provider.Puller); ok {
-				command.logger.Info("pulling available models for provider", "provider", p.ID())
-				for _, modelID := range command.ModelIDsWithAttributes {
-					if !strings.HasPrefix(modelID, p.ID()) { // TODO Move this into `NewModel` to validate that a model belongs to a provider.
-						panic(fmt.Errorf("model %s does not belong to provider %s", modelID, p.ID()))
-					}
-
-					if err := puller.Pull(command.logger, modelID); err != nil {
-						command.logger.Panicf("ERROR: could not pull model %q: %s", modelID, err)
-					}
-				}
-			}
-
+		// Gather models.
+		models := map[string]model.Model{}
+		for _, p := range providers {
 			ms, err := p.Models()
 			if err != nil {
 				command.logger.Panicf("ERROR: could not query models for provider %q: %s", p.ID(), err)
 			}
-
 			for _, m := range ms {
 				models[m.ID()] = m
-				evaluationContext.ProviderForModel[m] = p
 				evaluationConfiguration.Models.Available = append(evaluationConfiguration.Models.Available, m.ID())
+
+				if len(command.ModelIDsWithProviderAndAttributes) == 0 {
+					command.ModelIDsWithProviderAndAttributes = append(command.ModelIDsWithProviderAndAttributes, p.ID()+provider.ProviderModelSeparator+m.ID())
+				}
 			}
 		}
 		modelIDs := maps.Keys(models)
 		sort.Strings(modelIDs)
-		if len(command.ModelIDsWithAttributes) == 0 {
-			command.ModelIDsWithAttributes = modelIDs
-		} else {
-			for _, modelID := range command.ModelIDsWithAttributes {
-				if _, ok := models[modelID]; !ok {
-					command.logger.Panicf("ERROR: model %s does not exist. Valid models are: %s", modelID, strings.Join(modelIDs, ", "))
+		sort.Strings(command.ModelIDsWithProviderAndAttributes)
+
+		// Check and initialize models.
+		evaluationContext.ProviderForModel = map[model.Model]provider.Provider{}
+		for _, modelIDsWithProviderAndAttributes := range command.ModelIDsWithProviderAndAttributes {
+			command.logger.Info("selecting model", "model", modelIDsWithProviderAndAttributes)
+
+			providerID, modelIDsWithAttributes, ok := strings.Cut(modelIDsWithProviderAndAttributes, provider.ProviderModelSeparator)
```

Comment:
Please extract into a function similar to `ParseModelID`.

Reply:
This needs another round. Will do with another PR.